# Week 2 - Preprocessing, part 2

# 1. Lesson: None

# 2. Weekly graph question

The Storytelling With Data book mentions planning on a "Who, What, and How" for your data story.  Write down a possible Who, What, and How for your data, using the ideas in the book.

# 3. Homework - work with your own data

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

In [2]:
pip install ucimlrepo


In [3]:
from ucimlrepo import fetch_ucirepo

# fetch dataset
breast_cancer_wisconsin_diagnostic = fetch_ucirepo(id=17)

# data (as pandas dataframes)
X = breast_cancer_wisconsin_diagnostic.data.features
y = breast_cancer_wisconsin_diagnostic.data.targets

# metadata
print(breast_cancer_wisconsin_diagnostic.metadata)

# variable information
print(breast_cancer_wisconsin_diagnostic.variables)

{'uci_id': 17, 'name': 'Breast Cancer Wisconsin (Diagnostic)', 'repository_url': 'https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic', 'data_url': 'https://archive.ics.uci.edu/static/public/17/data.csv', 'abstract': 'Diagnostic Wisconsin Breast Cancer Database.', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 569, 'num_features': 30, 'feature_types': ['Real'], 'demographics': [], 'target_col': ['Diagnosis'], 'index_col': ['ID'], 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 1993, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C5DW2B', 'creators': ['William Wolberg', 'Olvi Mangasarian', 'Nick Street', 'W. Street'], 'intro_paper': {'ID': 230, 'type': 'NATIVE', 'title': 'Nuclear feature extraction for breast tumor diagnosis', 'authors': 'W. Street, W. Wolberg, O. Mangasarian', 'venue': 'Electronic imaging', 'year': 1993, 'journal': None, 'DOI': '1

In [2]:
datasource= "https://archive.ics.uci.edu/static/public/17/data.csv"

df_BreastCancer = pd.read_csv(datasource, sep=',', header=0) #loading breast cancer dataset
df_BreastCancer.sample(10, random_state=42) #printing random sample of 10 rows

ID  radius1  texture1  perimeter1   area1  smoothness1  \
204    87930    12.47     18.60       81.09   481.9      0.09965   
70    859575    18.94     21.31      123.60  1130.0      0.09009   
131     8670    15.46     19.48      101.70   748.9      0.10920   
431   907915    12.40     17.68       81.47   467.8      0.10540   
540   921385    11.54     14.44       74.65   402.9      0.09984   
567   927241    20.60     29.33      140.10  1265.0      0.11780   
369  9012000    22.01     21.90      147.20  1482.0      0.10630   
29    853201    17.57     15.05      115.00   955.1      0.09847   
81   8611161    13.34     15.86       86.49   520.0      0.10780   
477   911673    13.90     16.62       88.97   599.4      0.06828   

     compactness1  concavity1  concave_points1  symmetry1  ...  texture3  \
204       0.10580     0.08005          0.03821     0.1925  ...     24.64   
70        0.10290     0.10800          0.07951     0.1582  ...     26.58   
131       0.12230     0.14660          0.08087     0.1931  ...     26.00   
431       0.13160     0.07741          0.02799     0.1811  ...     22.91   
540       0.11200     0.06737          0.02594     0.1818  ...     19.68   
567       0.27700     0.35140          0.15200     0.2397  ...     39.42   
369       0.19540     0.24480          0.15010     0.1824  ...     25.80   
29        0.11570     0.09875          0.07953     0.1739  ...     19.52   
81        0.15350     0.11690          0.06987     0.1942  ...     23.19   
477       0.05319     0.02224          0.01339     0.1813  ...     21.80   

     perimeter3   area3  smoothness3  compactness3  concavity3  \
204       96.05   677.9      0.14260        0.2378      0.2671   
70       165.90  1866.0      0.11930        0.2336      0.2687   
131      124.90  1156.0      0.15460        0.2394      0.3791   
431       89.61   515.8      0.14500        0.2629      0.2403   
540       78.78   457.8      0.13450        0.2118      0.1797   
567      184.60  1821.0      0.16500        0.8681      0.9387   
369      195.00  2227.0      0.12940        0.3885      0.4756   
29       134.90  1227.0      0.12550        0.2812      0.2489   
81        96.66   614.9      0.15360        0.4791      0.4858   
477      101.20   718.9      0.09384        0.2006      0.1384   

     concave_points3  symmetry3  fractal_dimension3  Diagnosis  
204          0.10150     0.3014             0.08750          B  
70           0.17890     0.2551             0.06589          M  
131          0.15140     0.2837             0.08019          M  
431          0.07370     0.2556             0.09359          B  
540          0.06918     0.2329             0.08134          B  
567          0.26500     0.4087             0.12400          M  
369          0.24320     0.2741             0.08574          M  
29           0.14560     0.2756             0.07919          M  
81           0.17080     0.3527             0.10160          B  
477          0.06222     0.2679             0.07698          B  

[10 rows x 32 columns]

This week, you will do the same types of exercises as last week, but you should use your chosen datasets that someone in your class found last semester. (They likely will not be the particular datasets that you found yourself.)

### Here are some types of analysis you can do  Use Google, documentation, and ChatGPT to help you:

- Summarize the datasets using info() and describe()

- Are there any duplicate rows?

- Are there any duplicate values in a given column (when this would be inappropriate?)

- What are the mean, median, and mode of each column?

- Are there any missing or null values?

    - Do you want to fill in the missing value with a mean value?  A value of your choice?  Remove that row?

- Identify any other inconsistent data (e.g. someone seems to be taking an action before they are born.)

- Encode any categorical variables (e.g. with one-hot encoding.)

### Conclusions:

- Are the data usable?  If not, find some new data!

- Do you need to modify or correct the data in some way?

- Is there any class imbalance?  (Categories that have many more items than other categories).

In [3]:
#List of the columns
list[df_BreastCancer.columns]

list[Index(['ID', 'radius1', 'texture1', 'perimeter1', 'area1', 'smoothness1',
       'compactness1', 'concavity1', 'concave_points1', 'symmetry1',
       'fractal_dimension1', 'radius2', 'texture2', 'perimeter2', 'area2',
       'smoothness2', 'compactness2', 'concavity2', 'concave_points2',
       'symmetry2', 'fractal_dimension2', 'radius3', 'texture3', 'perimeter3',
       'area3', 'smoothness3', 'compactness3', 'concavity3', 'concave_points3',
       'symmetry3', 'fractal_dimension3', 'Diagnosis'],
      dtype='object')]

In [4]:
#Summarize the datasets using describe()
df_BreastCancer.describe()

ID     radius1    texture1  perimeter1        area1  \
count  5.690000e+02  569.000000  569.000000  569.000000   569.000000   
mean   3.037183e+07   14.127292   19.289649   91.969033   654.889104   
std    1.250206e+08    3.524049    4.301036   24.298981   351.914129   
min    8.670000e+03    6.981000    9.710000   43.790000   143.500000   
25%    8.692180e+05   11.700000   16.170000   75.170000   420.300000   
50%    9.060240e+05   13.370000   18.840000   86.240000   551.100000   
75%    8.813129e+06   15.780000   21.800000  104.100000   782.700000   
max    9.113205e+08   28.110000   39.280000  188.500000  2501.000000   

       smoothness1  compactness1  concavity1  concave_points1   symmetry1  \
count   569.000000    569.000000  569.000000       569.000000  569.000000   
mean      0.096360      0.104341    0.088799         0.048919    0.181162   
std       0.014064      0.052813    0.079720         0.038803    0.027414   
min       0.052630      0.019380    0.000000         0.000000    0.106000   
25%       0.086370      0.064920    0.029560         0.020310    0.161900   
50%       0.095870      0.092630    0.061540         0.033500    0.179200   
75%       0.105300      0.130400    0.130700         0.074000    0.195700   
max       0.163400      0.345400    0.426800         0.201200    0.304000   

       ...     radius3    texture3  perimeter3        area3  smoothness3  \
count  ...  569.000000  569.000000  569.000000   569.000000   569.000000   
mean   ...   16.269190   25.677223  107.261213   880.583128     0.132369   
std    ...    4.833242    6.146258   33.602542   569.356993     0.022832   
min    ...    7.930000   12.020000   50.410000   185.200000     0.071170   
25%    ...   13.010000   21.080000   84.110000   515.300000     0.116600   
50%    ...   14.970000   25.410000   97.660000   686.500000     0.131300   
75%    ...   18.790000   29.720000  125.400000  1084.000000     0.146000   
max    ...   36.040000   49.540000  251.200000  4254.000000     0.222600   

       compactness3  concavity3  concave_points3   symmetry3  \
count    569.000000  569.000000       569.000000  569.000000   
mean       0.254265    0.272188         0.114606    0.290076   
std        0.157336    0.208624         0.065732    0.061867   
min        0.027290    0.000000         0.000000    0.156500   
25%        0.147200    0.114500         0.064930    0.250400   
50%        0.211900    0.226700         0.099930    0.282200   
75%        0.339100    0.382900         0.161400    0.317900   
max        1.058000    1.252000         0.291000    0.663800   

       fractal_dimension3  
count          569.000000  
mean             0.083946  
std              0.018061  
min              0.055040  
25%              0.071460  
50%              0.080040  
75%              0.092080  
max              0.207500  

[8 rows x 31 columns]

In [23]:
#Summarize the datasets using info()
df_BreastCancer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  569 non-null    int64  
 1   radius1             569 non-null    float64
 2   texture1            569 non-null    float64
 3   perimeter1          569 non-null    float64
 4   area1               569 non-null    float64
 5   smoothness1         569 non-null    float64
 6   compactness1        569 non-null    float64
 7   concavity1          569 non-null    float64
 8   concave_points1     569 non-null    float64
 9   symmetry1           569 non-null    float64
 10  fractal_dimension1  569 non-null    float64
 11  radius2             569 non-null    float64
 12  texture2            569 non-null    float64
 13  perimeter2          569 non-null    float64
 14  area2               569 non-null    float64
 15  smoothness2         569 non-null    float64
 16  compactn

In [24]:
#Check duplicate rows:
df_BreastCancer.duplicated().any()
#No duplicate rows are found

np.False_

In [25]:
#Finding any duplicate values in the ID column
df_BreastCancer["ID"].duplicated().any()
#No duplicate IDs are found


np.False_

In [10]:
#Check whether there are any null values in the DataFrame
df_BreastCancer.isna().any().any()
#No null values are found

df_BreastCancer.isna().any()

ID                    False
radius1               False
texture1              False
perimeter1            False
area1                 False
smoothness1           False
compactness1          False
concavity1            False
concave_points1       False
symmetry1             False
fractal_dimension1    False
radius2               False
texture2              False
perimeter2            False
area2                 False
smoothness2           False
compactness2          False
concavity2            False
concave_points2       False
symmetry2             False
fractal_dimension2    False
radius3               False
texture3              False
perimeter3            False
area3                 False
smoothness3           False
compactness3          False
concavity3            False
concave_points3       False
symmetry3             False
fractal_dimension3    False
Diagnosis             False
dtype: bool

In [15]:
 #mean, median, and mode of each column
 #Mean
 df_BreastCancer_2=df_BreastCancer.drop("ID", axis=1)
 #print(df_BreastCancer_2)
 print("Mean")
 mean = df_BreastCancer_2.mean(numeric_only=True)
 print(mean)
 print("........................................")
 print("median")
 median=df_BreastCancer_2.median(numeric_only=True)
 print(median)
 print("........................................")
 print("mode")
 mode=df_BreastCancer_2.mode()
 print(mode.iloc[0])


     radius1  texture1  perimeter1   area1  smoothness1  compactness1  \
0      17.99     10.38      122.80  1001.0      0.11840       0.27760   
1      20.57     17.77      132.90  1326.0      0.08474       0.07864   
2      19.69     21.25      130.00  1203.0      0.10960       0.15990   
3      11.42     20.38       77.58   386.1      0.14250       0.28390   
4      20.29     14.34      135.10  1297.0      0.10030       0.13280   
..       ...       ...         ...     ...          ...           ...   
564    21.56     22.39      142.00  1479.0      0.11100       0.11590   
565    20.13     28.25      131.20  1261.0      0.09780       0.10340   
566    16.60     28.08      108.30   858.1      0.08455       0.10230   
567    20.60     29.33      140.10  1265.0      0.11780       0.27700   
568     7.76     24.54       47.92   181.0      0.05263       0.04362   

     concavity1  concave_points1  symmetry1  fractal_dimension1  ...  \
0       0.30010          0.14710     0.2419        

In [16]:
#Encode categorical variable Diagnosis
one_hot = pd.get_dummies(df_BreastCancer["Diagnosis"])
df2 = df_BreastCancer.join(one_hot)
df2

ID  radius1  texture1  perimeter1   area1  smoothness1  \
0      842302    17.99     10.38      122.80  1001.0      0.11840   
1      842517    20.57     17.77      132.90  1326.0      0.08474   
2    84300903    19.69     21.25      130.00  1203.0      0.10960   
3    84348301    11.42     20.38       77.58   386.1      0.14250   
4    84358402    20.29     14.34      135.10  1297.0      0.10030   
..        ...      ...       ...         ...     ...          ...   
564    926424    21.56     22.39      142.00  1479.0      0.11100   
565    926682    20.13     28.25      131.20  1261.0      0.09780   
566    926954    16.60     28.08      108.30   858.1      0.08455   
567    927241    20.60     29.33      140.10  1265.0      0.11780   
568     92751     7.76     24.54       47.92   181.0      0.05263   

     compactness1  concavity1  concave_points1  symmetry1  ...   area3  \
0         0.27760     0.30010          0.14710     0.2419  ...  2019.0   
1         0.07864     0.08690          0.07017     0.1812  ...  1956.0   
2         0.15990     0.19740          0.12790     0.2069  ...  1709.0   
3         0.28390     0.24140          0.10520     0.2597  ...   567.7   
4         0.13280     0.19800          0.10430     0.1809  ...  1575.0   
..            ...         ...              ...        ...  ...     ...   
564       0.11590     0.24390          0.13890     0.1726  ...  2027.0   
565       0.10340     0.14400          0.09791     0.1752  ...  1731.0   
566       0.10230     0.09251          0.05302     0.1590  ...  1124.0   
567       0.27700     0.35140          0.15200     0.2397  ...  1821.0   
568       0.04362     0.00000          0.00000     0.1587  ...   268.6   

     smoothness3  compactness3  concavity3  concave_points3  symmetry3  \
0        0.16220       0.66560      0.7119           0.2654     0.4601   
1        0.12380       0.18660      0.2416           0.1860     0.2750   
2        0.14440       0.42450      0.4504           0.2430     0.3613   
3        0.20980       0.86630      0.6869           0.2575     0.6638   
4        0.13740       0.20500      0.4000           0.1625     0.2364   
..           ...           ...         ...              ...        ...   
564      0.14100       0.21130      0.4107           0.2216     0.2060   
565      0.11660       0.19220      0.3215           0.1628     0.2572   
566      0.11390       0.30940      0.3403           0.1418     0.2218   
567      0.16500       0.86810      0.9387           0.2650     0.4087   
568      0.08996       0.06444      0.0000           0.0000     0.2871   

     fractal_dimension3  Diagnosis      B      M  
0               0.11890          M  False   True  
1               0.08902          M  False   True  
2               0.08758          M  False   True  
3               0.17300          M  False   True  
4               0.07678          M  False   True  
..                  ...        ...    ...    ...  
564             0.07115          M  False   True  
565             0.06637          M  False   True  
566             0.07820          M  False   True  
567             0.12400          M  False   True  
568             0.07039          B   True  False  

[569 rows x 34 columns]

In [17]:
#Check value of "Diagnosis" column
print(df_BreastCancer["Diagnosis"].value_counts(normalize=True)*100)

Diagnosis
B    62.741652
M    37.258348
Name: proportion, dtype: float64


**Conclusion**: The breast cancer dataset is a clean dataset with no missing value and no duplicates records. I do not need to modify any features. This dataset doesn’t have equal number of data of each class from the categorical column dignosis(B, M). From the analysis we can see benign case is 62.7% and malignant case is 37.3% out of 569 instances. Benign class is notably(almost twice) larger than malignant class. The classes are imbalance.

# 4. Storytelling With Data graph

Just like last week: choose any graph in the Introduction of Storytelling With Data (p. 1-17). Use matplotlib to reproduce it in a rough way. I don't expect you to spend an enormous amount of time on this; I understand that you likely will not have time to re-create every feature of the graph. However, if you're excited about learning to use matplotlib, this is a good way to do that. You don't have to duplicate the exact values on the graph; just the same rough shape will be enough.  If you don't feel comfortable using matplotlib yet, do the best you can and write down what you tried or what Google searches you did to find the answers.